src/ndarray/ndarray.cc

```diff
@@ -246,16 +246,20 @@ void TernaryOp(const NDArray &lhs,
 }
 
 /*!
- * \brief run a binary operation
+ * \brief Performs some preparation required to apply binary operators.
+ * Checks context and shape of ndarrays, allocates space for output
+ * and prepares const variables for engine
  * \param lhs left operand
  * \param rhs right operand
  * \param out the output ndarray
  * \param binary_op the real
+ * \param const_vars variables which are not changed by the binary operators
  */
 template<typename OP>
-void BinaryOp(const NDArray &lhs,
-              const NDArray &rhs,
-              NDArray *out) {
+void BinaryOpPrepare(const NDArray &lhs,
+                     const NDArray &rhs,
+                     NDArray *out,
+                     std::vector<Engine::VarHandle> const_vars) {
```

Comment:
`const std::vector<Engine::VarHandle>&` ?

Reply:
I made it a reference, but it should not be const, because it modifies the vector

Reply:
Okay changed to pointer as per code style guidelines (and lint)